mmaction/models/common/transformer.py

```diff
@@ -0,0 +1,209 @@
+import torch
+import torch.nn as nn
+from einops import rearrange
+from mmcv.cnn import build_norm_layer, constant_init
+from mmcv.cnn.bricks.registry import ATTENTION, FEEDFORWARD_NETWORK
+from mmcv.cnn.bricks.transformer import FFN, build_dropout
+from mmcv.runner.base_module import BaseModule
+
+
+@ATTENTION.register_module()
+class DividedTemporalAttentionWithNorm(BaseModule):
+    """Temporal Attention in Divided Space Time Attention.
+
+    Args:
+        embed_dims (int): Dimensions of embedding.
+        num_heads (int): Number of parallel attention heads in
+            TransformerCoder.
+        num_frames (int): Number of frames in the video.
+        attn_drop (float): A Dropout layer on attn_output_weights. Defaults to
+            0..
+        proj_drop (float): A Dropout layer after `nn.MultiheadAttention`.
+            Defaults to 0..
+        dropout_layer (dict): The dropout_layer used when adding the shortcut.
+            Defaults to `dict(type='DropPath', drop_prob=0.1)`.
+        norm_cfg (dict): Config dict for normalization layer. Defaults to
+            `dict(type='LN')`.
+        init_cfg (dict | None): The Config for initialization. Defaults to
+            None.
+    """
+
+    def __init__(self,
+                 embed_dims,
+                 num_heads,
+                 num_frames,
+                 attn_drop=0.,
+                 proj_drop=0.,
+                 dropout_layer=dict(type='DropPath', drop_prob=0.1),
+                 norm_cfg=dict(type='LN'),
+                 init_cfg=None,
+                 **kwargs):
+        super().__init__(init_cfg)
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        self.num_frames = num_frames
+        self.norm = build_norm_layer(norm_cfg, self.embed_dims)[1]
+        self.attn = nn.MultiheadAttention(embed_dims, num_heads, attn_drop,
```

Comment:
Perhaps you need to use MultiheadAttention from mmcv since it takes the arg `batch_first`

Reply:
Yes, it should be turned into mmcv MultiheadAttention, pytorch 1.9 includes arg but previous versions not.